2D and 3D Audio engine

AtlantiaKing opened the pull request:
### FEATURES:
- Load sounds in sync or async
- Play sounds both 2d and 3D
- Pause, resume, mute and unmute channels
- PauseAll, ResumeAll, MuteAll and UnmuteAll
- SetVolume on channels
- Check if a sound is loaded
- Check if a channel is playing

### TODO in main engine:
- Needs ServiceLocator
- Needs to be updated in the game loop

AtlantiaKing committed:
Created AudioSystem interface

Added SoundData3D struct which holds spatial data for a 3D sound
Added AudioSystem interface

AtlantiaKing committed:
Added AudioEngine branch to cmake.yml

AtlantiaKing committed:
Added Services and Servicelocator

Added DefaultAudioSystem and FmodAudioSystem to the AudioEngine
Added ServiceLocator to base engine

AtlantiaKing committed:
Merge branch 'AudioEngine' of https://github.com/LeeVangraefschepe/LeapEngine into AudioEngine

AtlantiaKing committed:
FMOD initialization

Added Update method to the AudioSystem interface
Added functionality to the constructor, destructor and update method of the FmodAudioSystem

AtlantiaKing committed:
Added sync sound loading and 2D sounds

Added functionality to the LoadSound and PlaySound2D methods of the FmodAudioSystem class
Added FMODSound struct to the FmodAudioSystem class
Added testcode to the Leap class

AtlantiaKing committed:
Added pImpl idiom to FmodAudioSystem

AtlantiaKing committed:
Added async sound loading

Remove callback from the LoadSoundAsync method of the AudioSystem
Added IsValidSound method to the AudioSystem classes

AtlantiaKing committed:
Removed callbacks for PlaySound methods

AtlantiaKing committed:
Added runtime error on play fail

AtlantiaKing committed:
Added SetVolume2D

Fixed channels not being returned from the PlaySound methods of the AudioSystem classes
Renamed id parameter to channel where it is appropriate of the AudioSystem classes
Added functionality to the SetVolume2D method of the FmodAudioSystem

AtlantiaKing committed:
Fixed W4 errors in the DefaultAudioSystem

AtlantiaKing committed:
Added Pausing/Unpausing

Added functionality for Pause, Resume, PauseAll and ResumeAll methods of the FmodAudioSystem class
Added a list of playing channels to each sound
Modified Update method of the FmodAudioSystem to remove channels that are no longer playing

AtlantiaKing committed:
Moved all playing channel data to a struct

Added ChannelData struct which holds the id, paused state and volume
Modified FmodAudioSystem methods to use the ChannelData struct instead of the unordered_map

AtlantiaKing committed:
Added mute/unmute

Replaced volume with a muted state in the ChannelData
Added functionality for the Mute/Unmute/MuteAll/UnmuteAll methods of the FmodAudioSystem

AtlantiaKing committed:
Added basic 3D sounds

Increased minDistance of SoundData to 5 instead of 0
Added is3DSound paramater to the LoadSound methods of the AudioSystems
Modified the LoadSound methods of the FmodAudioSystem to be able to load 3D sounds
Added functionality to PlaySound3D/UpdateSound3D/UpdateListener3D methods of the FmodAudioSystem

AtlantiaKing committed:
Added velocity and rotation to 3D listener

Added parameters for velocity, forward vector and up vector to the UpdateListener3D method of the AudioSystems
Modified the UpdateListener3D method of the FmodAudioSystem to use the new parameters
Removed setting spread options when starting to play a 3D sound

AtlantiaKing committed:
Added IsPlaying method

Added IsPlaying method to AudioSystems
Added functionality to the IsPlaying method of the FmodAudioSystem

AtlantiaKing committed:
Reverted the LeapEngine to its previous state

Removed all data files
Removed ServiceLocator
Reverted Leap.cpp

AtlantiaKing committed:
Removed AudioEngine from build process

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                for (auto it{ begin(channels) }; it != end(channels);)
+                {
+                    // Retrieve the current channel
+                    FMOD::Channel* pChannel{};
```
Create channel outside for loops

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                        throw std::runtime_error("FMODAudioSystem Error: Can't retieve channel with this id");
+
+                    // Retrieve the playing state of this channel
+                    bool isPlaying{};
```
Create bool outside for loops

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                    {
+                        it = channels.erase(it);
+                    }
+                    else
```
Else statement not needed (use early out)

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                for (const auto& channelData : sound.channels)
+                {
+                    // Retrieve the current channel
+                    FMOD::Channel* pChannel{};
```
Create channel outside for loop

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                    if (channelData.paused) continue;
+
+                    // Retrieve the current channel
+                    FMOD::Channel* pChannel{};
```
Create channel outside for loop

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+        void Mute(int channel)
+        {
+            // Retrieve the current channel
+            FMOD::Channel* pChannel{};
```
Create channel outside for loop

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                for (const auto& channelData : sound.channels)
+                {
+                    // Retrieve the current channel
+                    FMOD::Channel* pChannel{};
```
Create channel outside for loop

LeeVangraefschepe commented on Audio/Systems/FmodAudioSystem.cpp:
```diff
@@ -0,0 +1,640 @@
+                    if (channelData.muted) continue;
+
+                    // Retrieve the current channel
+                    FMOD::Channel* pChannel{};
```
Create channel outside for loop

LeeVangraefschepe reviewed (approved):
Check the comments for optimization

AtlantiaKing committed:
For loop optimizations

Moved reusable variabeles out of for loops
Removed else statements where early out was possible

LeeVangraefschepe reviewed (approved):
Feedback has been implemented
